# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/clients/manager/ClientHandleParent.cpp
# Commit: 6125bb02952a
# Full Hash: 6125bb02952aff71c0d757e238a7752101b2d6c2
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2019-09-03 04:06:20
# Regressor Bug: 1556489
# File Overlap Count: 1
# Description:
#   Bug 1556489 - P23 - Only create new ClientSource objects in the content, but use the recreated info from redirects in the parent. r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D42260
# ==============================================================================

diff -r 06529d12b589 -r 6125bb02952a dom/clients/manager/ClientHandleParent.cpp
--- a/dom/clients/manager/ClientHandleParent.cpp	Mon Aug 26 12:18:57 2019 +1000
+++ b/dom/clients/manager/ClientHandleParent.cpp	Tue Aug 27 08:53:37 2019 +1000
@@ -26,6 +26,12 @@
   if (mSource) {
     mSource->DetachHandle(this);
     mSource = nullptr;
+  } else {
+    mService->StopWaitingForSource(this, mClientId);
+  }
+
+  if (mSourcePromise) {
+    mSourcePromise->Reject(NS_ERROR_FAILURE, __func__);
   }
 }
 
@@ -53,9 +59,11 @@
 ClientHandleParent::~ClientHandleParent() { MOZ_DIAGNOSTIC_ASSERT(!mSource); }
 
 void ClientHandleParent::Init(const IPCClientInfo& aClientInfo) {
+  mClientId = aClientInfo.id();
+  mPrincipalInfo = aClientInfo.principalInfo();
   mSource = mService->FindSource(aClientInfo.id(), aClientInfo.principalInfo());
   if (!mSource) {
-    Unused << Send__delete__(this);
+    mService->WaitForSource(this, aClientInfo.id());
     return;
   }
 
@@ -64,5 +72,32 @@
 
 ClientSourceParent* ClientHandleParent::GetSource() const { return mSource; }
 
+RefPtr<SourcePromise> ClientHandleParent::EnsureSource() {
+  if (mSource) {
+    return SourcePromise::CreateAndResolve(mSource, __func__);
+  }
+
+  mSourcePromise = new SourcePromise::Private(__func__);
+  return mSourcePromise;
+}
+
+void ClientHandleParent::FoundSource(ClientSourceParent* aSource) {
+  MOZ_ASSERT(aSource->Info().Id() == mClientId);
+  if (!ClientMatchPrincipalInfo(aSource->Info().PrincipalInfo(),
+                                mPrincipalInfo)) {
+    if (mSourcePromise) {
+      mSourcePromise->Reject(NS_ERROR_FAILURE, __func__);
+    }
+    Unused << Send__delete__(this);
+    return;
+  }
+
+  mSource = aSource;
+  mSource->AttachHandle(this);
+  if (mSourcePromise) {
+    mSourcePromise->Resolve(aSource, __func__);
+  }
+}
+
 }  // namespace dom
 }  // namespace mozilla